### PR TITLE
Add lua functions for Ion Cannon and Nuclear Missile

### PR DESCRIPTION
**Pull request by MustaphaTR**
Testcases adds a nuke to RA shellmap and ion cannon firing on Nod9.

### CODE REVIEW

**Comment by abcdefg30** [comment]
This was once rejected at #11030.

**Comment by MustaphaTR** [comment]
I think this worths to add to be later removed, because i don't think normal weapons/armaments for nuke and ion cannon will happen soon.

**Comment by GraionDilach** [comment]
TBH, the AttackOrderPower implementation and setup are terrible, so and AS-FireArmamentPower just tops it in many areas.

File: OpenRA.Mods.Cnc/Traits/SupportPowers/IonCannonPower.cs [hunk 1]
```diff
@@ -74,19 +74,24 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
 		{
 			base.Activate(self, order, manager);
 
+			SendIonCannon(self, order.TargetLocation);
+		}
+
+		public void SendIonCannon(Actor self, CPos targetLocation)
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/SupportPowers/IonCannonPower.cs]: Can we please call this `Activate(Actor self, CPos targetLocation)` instead?

File: OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs [hunk 2]
```diff
@@ -108,10 +108,14 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
 			base.Activate(self, order, manager);
 			PlayLaunchSounds();
 
+			SendNuke(self, self.World.Map.CenterOfCell(order.TargetLocation));
+		}
+
+		public void SendNuke(Actor self, WPos targetPosition)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs]: Can we please call this `Activate(Actor self, CPos targetLocation)` instead?

File: OpenRA.Mods.Cnc/Scripting/Properties/IonCannonProperties.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		readonly IonCannonPower icp;
+
+		public IonCannonProperties(ScriptContext context, Actor self)
+			: base(context, self)
+		{
+			icp = self.TraitsImplementing<IonCannonPower>().First();
+		}
+
+		[Desc("Activate the actor's IonCannonPower.")]
+		public void SendIonCannon(CPos target)
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Scripting/Properties/IonCannonProperties.cs]: Can we please call this `ActivateIonCannon` instead?

File: OpenRA.Mods.Common/Scripting/Properties/NukeProperties.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		readonly NukePower np;
+
+		public NukeProperties(ScriptContext context, Actor self)
+			: base(context, self)
+		{
+			np = self.TraitsImplementing<NukePower>().First();
+		}
+
+		[Desc("Activate the actor's NukePower.")]
+		public void SendNuke(CPos target)
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/NukeProperties.cs]: Can we please call this `ActivateNukePower` instead?

**Review by pchote — COMMENTED** [review]
It seems unlikely that these powers will be removed any time soon, so we may as well implement support for mappers in the meantime.

**Comment by MustaphaTR** [comment]
Updated.

**Review by GraionDilach — DISMISSED** [review]
:+1:

File: mods/ra/maps/desert-shellmap/desert-shellmap.lua [hunk 5]
```diff
@@ -176,6 +176,10 @@ WorldLoaded = function()
 	soviets = Player.GetPlayer("Soviets")
 	viewportOrigin = Camera.Position
 
+	Trigger.AfterDelay(DateTime.Seconds(10), function()
+		MissileSilo.SendNuke(Actor303.Location)
```
Comment (Mailaender) [inline mods/ra/maps/desert-shellmap/desert-shellmap.lua]:
```
Fatal Lua Error: [string "BindingSupport.lua"]:30: Actor 'mslo' does not define a property 'SendNuke'
  at OpenRA.Scripting.ScriptContext.FatalError (System.String message) [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Mods.Common.Scripting.TriggerGlobal+<AfterDelay>c__AnonStorey0.<>m__0 () [0x00000] in <8519c22846ec4f12b51f5d7177d1eefa>:0 
  at OpenRA.Effects.DelayedAction.<Tick>m__0 (OpenRA.World w) [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.World.Tick () [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Game.LogicTick () [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Game.Loop () [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Game.Run () [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00000] in <4dbc9595f4ec4242bf7648a430506bc4>:0
```

Comment (Mailaender) [reply mods/ra/maps/desert-shellmap/desert-shellmap.lua]: Changing it to `ActivateNukePower` fixes it.

Comment (MustaphaTR) [reply mods/ra/maps/desert-shellmap/desert-shellmap.lua]: I renamed the properties as pchote wanted but forgot to update testcase. Shouldn't be much problem as it is just testcase. Can you test by manually updating it?

**Comment by Mailaender** [comment]
Works as promised. :heavy_check_mark: Waiting for the testcases to be removed.

**Comment by MustaphaTR** [comment]
Updated, removed the testcase.

**Review by Mailaender — APPROVED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bc677865d108ec80bc1044d43d4d46c68c8b0781...bdba9652d31313ce6758ddb7edb0eef0a0bda5419)

**Pull request merged**